server/src/main/java/io/crate/replication/logical/LogicalReplicationService.java

```diff
@@ -301,44 +344,125 @@ public void verifyTablesDoNotExist(String subscriptionName,
+                            msg = String.format(Locale.ENGLISH,
+                                                "Restoring the subscription relations failed partially. Failed to restore %d/%d shards",
+                                                restoreInfo.failedShards(),
+                                                restoreInfo.totalShards());
+                        }
+                        updateSubscriptionState(
+                            subscriptionName,
+                            Subscription.State.FAILED,
+                            msg
+                        ).whenComplete(
+                            (ignore, ignoredErr) -> future.complete(false)
+                        );
+                    }
                 } else {
-                    assert restoreInfo.failedShards() > 0 : "Some failed shards are expected";
-                    LOGGER.error("Failed to restore {}/{} shards",
-                                 restoreInfo.failedShards(),
-                                 restoreInfo.totalShards());
-                    return false;
+                    updateSubscriptionState(
+                        subscriptionName,
+                        Subscription.State.FAILED,
+                        "Error while initial restoring the subscription relations"
+                    ).whenComplete(
+                        (ignore, ignoredErr) -> future.completeExceptionally(err)
+                    );
                 }
-
             }
         );
-        clusterService.addListener(new RestoreClusterStateListener(clusterService, response, future));
+        clusterService.addListener(new RestoreClusterStateListener(clusterService, response, restoreFuture));
+        return future;
+    }
+
+    private CompletableFuture<Boolean> updateSubscriptionState(
+        final String subscriptionName,
+        final Subscription.State newState,
+        @Nullable final String failureReason) {
+        var oldSubscription = subscriptions().get(subscriptionName);
+        if (oldSubscription == null) {
+            return CompletableFuture.completedFuture(false);
+        }
+        HashMap<RelationName, Subscription.RelationState> relations = new HashMap<>();
+        for (var entry : oldSubscription.relations().entrySet()) {
+            relations.put(entry.getKey(), new Subscription.RelationState(newState, failureReason));
+        }
+        var newSubscription = new Subscription(
+            oldSubscription.owner(),
+            oldSubscription.connectionInfo(),
+            oldSubscription.publications(),
+            oldSubscription.settings(),
+            relations
+        );
+        var request = new UpdateSubscriptionAction.Request(subscriptionName, newSubscription);
+        var future = new FutureActionListener<>(AcknowledgedResponse::isAcknowledged);
+        client.execute(UpdateSubscriptionAction.INSTANCE, request, future);
```

Comment:
Might be worth to already add this `client.execute(actionType, request) -> CompletableFuture` overload to the client.

Reply:
Unfortunately this overload already exists returning an `ActionFuture`: `client.execute(actionType, request) -> ActionFuture`